### PR TITLE
OpenAPI downloads protobuf rather than Json

### PR DESCRIPTION
**Pull request by apelisse**
**What this PR does / why we need it**: 
The current implementation of the OpenAPI getter fetches the swagger in a Json format from the apiserver. The Json file is big (~1.7mb), which means that it takes a long time to download, and then a long time to parse. Because that is going to be needed on each `kubectl` run later, we want this to be as fast as possible.

The apiserver has been modified to be able to return a protobuf version of the swagger, which this patch intends to use.

Note that there is currently no piece of code that exists that allows us to go from the protobuf version of the file, back into Json and/or `spec.Swagger`. Because the protobuf is not very different (but significantly different enough that it can't be translated), I've updated the code to use `openapi_v2.Document` (the protobuf type) everywhere rather than `spec.Swagger`. The behavior should be identical though.

There are more changes that are coming in follow-up pull-requests: using the gzip version (also provided by the new apiserver) to even further reduce the size of the downloaded content, and use the HTTP Etag cache mechanism to completely get rid of recurrent fetch requests. I'm currently working on these two features.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: partly #38637

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by mengqiy** [comment]
@apelisse Why we switch from `github.com/go-openapi/spec` to `github.com/googleapis/gnostic/OpenAPIv2`?

File: pkg/kubectl/cmd/util/openapi/openapi.go [hunk 1]
```diff
@@ -202,13 +228,13 @@ func (o *Resources) parseDefinition(name string, s spec.Schema) Kind {
 	if o.isPrimitive(s) {
 		value.PrimitiveType = o.getTypeNameForField(s)
 	}
-	for fieldname, property := range s.Properties {
-		value.Fields[fieldname] = o.parseField(property)
+	for _, ns := range s.Properties.GetAdditionalProperties() {
```
Comment (mengqiy) [inline pkg/kubectl/cmd/util/openapi/openapi.go]: Is it safer to use `s.GetProperties().GetAdditionalProperties()`?

File: pkg/kubectl/cmd/util/openapi/openapi.go [hunk 2]
```diff
@@ -266,96 +292,95 @@ func (o *Resources) isPrimitive(s spec.Schema) bool {
...
-func (o *Resources) getTypeNameForField(s spec.Schema) string {
+func (o *Resources) getTypeNameForField(s *openapi_v2.Schema) string {
 	// Get the reference for complex types
 	if o.isDefinitionReference(s) {
 		return o.nameForDefinitionField(s)
 	}
 	// Recurse if type is array
 	if o.isArray(s) {
-		return fmt.Sprintf("%s array", o.getTypeNameForField(*s.Items.Schema))
+		return fmt.Sprintf("%s array", o.getTypeNameForField(s.GetItems().GetSchema()[0]))
```
Comment (mengqiy) [inline pkg/kubectl/cmd/util/openapi/openapi.go]: Is it guaranteed to have at least one item?

File: pkg/kubectl/cmd/util/openapi/openapi.go [hunk 3]
```diff
@@ -266,96 +292,95 @@ func (o *Resources) isPrimitive(s spec.Schema) bool {
...
 // getElementType returns the type of an element for arrays
 // returns an error if s is not an array.
-func (o *Resources) getElementType(s spec.Schema) (spec.Schema, error) {
+func (o *Resources) getElementType(s *openapi_v2.Schema) (*openapi_v2.Schema, error) {
 	if !o.isArray(s) {
-		return spec.Schema{}, fmt.Errorf("%v is not an array type", s.Type)
+		return &openapi_v2.Schema{}, fmt.Errorf("%v is not an array type", o.getTypeNameForField(s))
 	}
-	return *s.Items.Schema, nil
+	return s.GetItems().GetSchema()[0], nil
```
Comment (mengqiy) [inline pkg/kubectl/cmd/util/openapi/openapi.go]: Same question here. Should we check the length of the schema slice?

File: pkg/kubectl/cmd/util/openapi/openapi.go [hunk 4]
```diff
@@ -266,96 +292,95 @@ func (o *Resources) isPrimitive(s spec.Schema) bool {
...
 // extensions
-// Expected format for s.Extensions: map[string][]map[string]string
 // map[x-kubernetes-group-version-kind:[map[Group:authentication.k8s.io Version:v1 Kind:TokenReview]]]
-func (o *Resources) getGroupVersionKind(s spec.Schema) (schema.GroupVersionKind, error) {
+func (o *Resources) getGroupVersionKind(s *openapi_v2.Schema) (schema.GroupVersionKind, error) {
 	empty := schema.GroupVersionKind{}
 
 	// Get the extensions
-	extList, f := s.Extensions[groupVersionKindExtensionKey]
+	extList, f := vendorExtensionToMap(s.GetVendorExtension())[groupVersionKindExtensionKey]
```
Comment (mengqiy) [inline pkg/kubectl/cmd/util/openapi/openapi.go]: Maybe save the result of `vendorExtensionToMap(s.GetVendorExtension())` to avoid process this again below.

File: pkg/kubectl/cmd/util/openapi/openapi.go [hunk 5]
```diff
@@ -366,9 +391,9 @@ func (o *Resources) getGroupVersionKind(s spec.Schema) (schema.GroupVersionKind,
 	gvk := extListCasted[0]
 
 	// Expect a empty of a map with 3 entries
-	gvkMap, ok := gvk.(map[string]interface{})
+	gvkMap, ok := gvk.(map[interface{}]interface{})
```
Comment (mengqiy) [inline pkg/kubectl/cmd/util/openapi/openapi.go]: Interesting. It is not a `map[string]interface{}` any more?

File: pkg/kubectl/cmd/util/openapi/openapi.go [hunk 6]
```diff
@@ -20,8 +20,10 @@ import (
 	"fmt"
 	"strings"
 
-	"github.com/go-openapi/spec"
+	yaml "gopkg.in/yaml.v2"
```
Comment (mengqiy) [inline pkg/kubectl/cmd/util/openapi/openapi.go]: Not sure if we should use this or `github.com/ghodss/yaml`?

File: staging/src/k8s.io/client-go/discovery/discovery_client.go [hunk 7]
```diff
@@ -377,17 +377,17 @@ func (d *DiscoveryClient) SwaggerSchema(version schema.GroupVersion) (*swagger.A
 
 // OpenAPISchema fetches the open api schema using a rest client and parses the json.
 // Warning: this is very expensive (~1.2s)
-func (d *DiscoveryClient) OpenAPISchema() (*spec.Swagger, error) {
-	data, err := d.restClient.Get().AbsPath("/swagger.json").Do().Raw()
+func (d *DiscoveryClient) OpenAPISchema() (*openapi_v2.Document, error) {
+	data, err := d.restClient.Get().AbsPath("/swagger-2.0.0.pb-v1").Do().Raw()
```
Comment (mengqiy) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go]: Are you sure about the name.
I found `/swagger.pb` in https://docs.google.com/document/d/1EcnvLRqP9-5mqQ-kghfnB3GtB8eIrkVQjn1B16NRrr0/edit#heading=h.xgjl2srtytjt

File: pkg/kubectl/cmd/util/openapi/openapi.go (hunk 5 above)
Comment (apelisse) [inline pkg/kubectl/cmd/util/openapi/openapi.go]: Yeah, seems like the new type uses a different format.

File: pkg/kubectl/cmd/util/openapi/openapi.go (hunk 6 above)
Comment (mengqiy) [inline pkg/kubectl/cmd/util/openapi/openapi.go]: @fabianofranz Do you know what yaml pkg should we use?

File: pkg/kubectl/cmd/util/openapi/openapi.go [hunk 8]
```diff
@@ -98,7 +100,7 @@ type Kind struct {
 	PrimitiveType string
 
 	// Extensions are openapi extensions for the object definition.
```
Comment (mengqiy) [inline pkg/kubectl/cmd/util/openapi/openapi.go]: Maybe not. They are the same the type under the cover.

File: pkg/kubectl/cmd/util/cached_discovery.go [hunk 9]
```diff
@@ -25,8 +25,8 @@ import (
 	"time"
 
 	"github.com/emicklei/go-restful-swagger12"
-	"github.com/go-openapi/spec"
 	"github.com/golang/glog"
+	openapi_v2 "github.com/googleapis/gnostic/OpenAPIv2"
```
Comment (droot) [inline pkg/kubectl/cmd/util/cached_discovery.go]: wondering if we have any convention for package import names ? using "_" is a standard thing (I see metav1 below) ?

File: pkg/kubectl/cmd/util/openapi/openapi.go [hunk 10]
```diff
@@ -130,21 +132,45 @@ type Type struct {
...
 	o := &Resources{
 		GroupVersionKindToName: map[schema.GroupVersionKind]string{},
 		NameToDefinition:       map[string]Kind{},
 	}
 	// Parse and index definitions by name
-	for name, d := range s.Definitions {
-		definition := o.parseDefinition(name, d)
-		o.NameToDefinition[name] = definition
+	for _, ns := range doc.GetDefinitions().GetAdditionalProperties() {
+		definition := o.parseDefinition(ns.Name, ns.Value)
```
Comment (droot) [inline pkg/kubectl/cmd/util/openapi/openapi.go]: does ns have Getters for name and value. If yes, then we should use those to be consistent.

**Review by droot — COMMENTED** [review]
looks good to me, a few minor comments.

File: pkg/kubectl/cmd/util/openapi/openapi.go (hunk 10 above)
Comment (apelisse) [inline pkg/kubectl/cmd/util/openapi/openapi.go]: Yes, good catch!

File: pkg/kubectl/cmd/util/cached_discovery.go (hunk 9 above)
Comment (apelisse) [inline pkg/kubectl/cmd/util/cached_discovery.go]: This is mostly because my `goimports` was broken. I've fixed it and I'm using the default name now (which happens to also openapi_v2 ;-)

File: pkg/kubectl/cmd/util/openapi/openapi.go (hunk 6 above)
Comment (apelisse) [inline pkg/kubectl/cmd/util/openapi/openapi.go]: Actually we don't have a choice. I use a feature from `gopkg.in/yaml.v2`

**Comment by apelisse** [comment]
Addressed! Please take a look!

**Comment by mengqiy** [comment]
@k8s-bot pull-kubernetes-unit test this
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by mengqiy** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by mbohlool** [comment]
> apiserver. The Json file is big (~330mb), which means that it takes a long time to download, and then 

330mb? it is 1.7MB. right?

But the point is valid that it takes much longer to unmarshal. Some numbers here: https://docs.google.com/document/d/1EcnvLRqP9-5mqQ-kghfnB3GtB8eIrkVQjn1B16NRrr0

**Comment by apelisse** [comment]
> 330mb? it is 1.7MB. right?

I have no idea how I made up that number :D. I'll update the description.

**Comment by apelisse** [comment]
@pwittrock Please take a look!

**Comment by apelisse** [comment]
I've added a new commit to download the gzipped protobuf.

**Comment by pwittrock** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

File: staging/src/k8s.io/client-go/discovery/discovery_client.go [hunk 11]
```diff
@@ -375,19 +377,27 @@ func (d *DiscoveryClient) SwaggerSchema(version schema.GroupVersion) (*swagger.A
 	return &schema, nil
 }
 
-// OpenAPISchema fetches the open api schema using a rest client and parses the json.
-// Warning: this is very expensive (~1.2s)
-func (d *DiscoveryClient) OpenAPISchema() (*spec.Swagger, error) {
-	data, err := d.restClient.Get().AbsPath("/swagger.json").Do().Raw()
+// OpenAPISchema fetches the open api schema using a rest client and parses the proto.
+func (d *DiscoveryClient) OpenAPISchema() (*openapi_v2.Document, error) {
+	stream, err := d.restClient.Get().AbsPath("/swagger-2.0.0.pb-v1.gz").Stream()
```
Comment (deads2k) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go]: This looks off to me.  Why wouldn't you request a gzipped encoding of the normal protobuf struct?

Comment (apelisse) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go]: Because the server doesn't support it

Comment (deads2k) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go]: > Because the server doesn't support it

I'm pretty sure it merged recently here: https://github.com/kubernetes/kubernetes/pull/46966

Comment (apelisse) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go]: This code is targeting [this end-point](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/openapi/openapi_handler.go#L47), is it related?

**Comment by deads2k** [comment]
> This code is targeting this end-point, is it related?

The handler was supposed to be written as a "normal" http handler wrapper.  Can you try wrapping it?  I'd rather not enshrine a temporary condition in the API.

**Comment by apelisse** [comment]
Yeah I'll look at it

**Comment by apelisse** [comment]
OK I've just removed that part. Go will take care of that part if the server supports the gzip encoding.

**Comment by apelisse** [comment]
Can you PTAL @deads2k

**Comment by mbohlool** [comment]
If it is not too much trouble, can you separate generated/godep commits from others?

File: staging/src/k8s.io/client-go/discovery/discovery_client.go [hunk 12]
```diff
@@ -375,19 +375,18 @@ func (d *DiscoveryClient) SwaggerSchema(version schema.GroupVersion) (*swagger.A
 	return &schema, nil
 }
 
-// OpenAPISchema fetches the open api schema using a rest client and parses the json.
-// Warning: this is very expensive (~1.2s)
-func (d *DiscoveryClient) OpenAPISchema() (*spec.Swagger, error) {
-	data, err := d.restClient.Get().AbsPath("/swagger.json").Do().Raw()
+// OpenAPISchema fetches the open api schema using a rest client and parses the proto.
+func (d *DiscoveryClient) OpenAPISchema() (*openapi_v2.Document, error) {
+	data, err := d.restClient.Get().AbsPath("/swagger-2.0.0.pb-v1").Do().Raw()
```
Comment (deads2k) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go]: Ok, this now looks good apiserver-wise, but client-side it seems like this will break when run against old servers.  Don't you need a fallback to the old swagger.json?

Comment (apelisse) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go]: It shouldn't. 1.7 server already supports that end-point, and this code is meant for 1.8.

Comment (deads2k) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go]: > It shouldn't. 1.7 server already supports that end-point, and this code is meant for 1.8 client.

I rather not have this be the thing that breaks generic commands (like create) from working.  If it returns an error, will kubectl still work?  If not, at least guarding it sufficiently to keep that working seems reasonable even if client side validation doesn't work.

Comment (pwittrock) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go]: If --validate=false, we shouldn't block on getting the openapi.  If --validate=true, we should fail if we can't get the openapi to validate.

Comment (apelisse) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go]: IOW, This is fine for now as this code is not even used for validation. The only way to actually exercise this code today is to use `--experimental-use-openapi-print-columns`.

Comment (apelisse) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go]: The plan is to change validation to use this code, but even then, one will be able to go back to the old swagger validation with a flag.

Comment (deads2k) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go]: > The plan is to change validation to use this code, but even then, one will be able to go back to the old swagger validation with a flag

It still seems capricious and unnecessary to break.  We are just talking about keeping the old code in the `if err != nil` block, right?



> If --validate=true, we should fail if we can't get the openapi to validate.

The only reason you "can't" is for want of a few lines right here.

Comment (apelisse) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go]: > It still seems capricious and unnecessary to break. We are just talking about keeping the old code in the if err != nil block, right?

I would definitely not argue if that was it :-). 

We used to parse the json file into a `spec.Schema`, while the protobuf is parsed into a `openapi_v2.Document`. While there is code to go from `spec.Schema` to `openapi_v2.Document` (that code is used by the API-server to serve both json and protobuf), there is no code to go the other way around. That's why this PR is also updating every `spec.Schema` types to `openapi_v2.Document`.

I could still download the json file and convert it to protobuf and be happy, but to be honest, the json is 1.7mb (17x bigger than the gzipped protobuf), and the previous versions of the server don't support http caching, so it would have to be downloaded every-time, making it pretty much unusable anyway. 

In other words, I don't think kubectl of any version should ever use openapi with a pre-1.7 server version.

Comment (deads2k) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go]: > In other words, I don't think kubectl of any version should ever use openapi with a pre-1.7 server version.

Given code to go from spec.Schema -> openapi_v2.Document, it seems like you could hit the old endpoint, read into spec.Schema (slow as it is), then convert that to the openapi_v2.Document to return.

I'm fine telling people they shouldn't use the feature against older servers.  I'm less fine straight up breaking them.  With `--validate=true` being the default, that's what would happen with a 1.8 kubectl and 1.6 server (which a client may not be able to update), right?  Openapi for validation is a goal.

Comment (apelisse) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go]: We could keep the code as it is, have it fail on validating with the openapi, and then automatically fallback on swagger. That would allow us even longer backward compatibility.

Comment (deads2k) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go]: > We could keep the code as it is, have it fail on validating with the openapi, and then automatically fallback on swagger. That would allow us even longer backward compatibility.

Sold.  Is it practical do in this pull?

Comment (apelisse) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go]: Nothing really uses this code so far, and I'm working on adding new features but it's too substantial to be added to this pr. I'll definitely add you as a reviewer/approver on pull-requests to come.

Comment (deads2k) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go]: > Nothing really uses this code so far, and I'm working on adding new features but it's too substantial to be added to this pr. I'll definitely add you as a reviewer/approver on pull-requests to come.

Ok.  Please open an issue for yourself about maintaining compatibility.

**Comment by deads2k** [comment]
I'm ok with an issue that you'll fix in 1.8 to maintain compatibility.

/approve
